aea/agent_loop.py

```diff
@@ -0,0 +1,184 @@
+        self.behaviour = behaviour
+        self._timerhandle = None
+
+    def _callback(self):
+        self._schedule_call()
+        self._call_act()
+
+    def _call_act(self):
+        self.behaviour.act()
+
+    def _schedule_call(self):
+        if self.behaviour._tick_interval is None:
+            return
+
+        if self._timerhandle is None:
+            ts = time.mktime(self.behaviour._start_at.timetuple())
+            delay = max(0, ts - time.time())
+            self._timerhandle = self._loop.call_later(delay, self._callback)
+        else:
+            self._timerhandle = self._loop.call_later(
+                self.behaviour._tick_interval, self._callback
+            )
+
+    def start(self):
+        if self._timerhandle:
+            return
+        self._schedule_call()
+
+    def stop(self):
+        if not self._timerhandle:
+            return
+
+        self._timerhandle.cancel()
+        self._timerhandle = None
+
+
+class AgentLoop:
+    def __init__(self, inbox, handlers=None, behaviours=None, loop=None):
+        self._handerls = set(handlers or [])
+        self._behaviours = set(behaviours or [])
+        self._inbox = inbox
+        self._loop = loop or asyncio.get_event_loop()
+        self._state = AsyncState()
+        self._registered_behaviours = {}
+
+    def _register_behaviour(self, behaviour):
+        scheuled_behaviour = ScheduledBehaviour(behaviour, self._loop)
+        self._registered_behaviours[behaviour] = scheuled_behaviour
+        scheuled_behaviour.start()
+
+    def _stop_all_behaviours(self):
+        for scheduled_behaviour in self._registered_behaviours.values():
+            scheduled_behaviour.stop()
+        self._registered_behaviours = {}
+
+    def _unregister_behaviour(self, behaviour):
+        scheduled_behaviour = self._registered_behaviours.pop(behaviour, None)
+        if not scheduled_behaviour:
+            return
+        scheduled_behaviour.stop()
+
+    def start(self):
+        self._state.set("started")
+        self._loop.run_until_complete(self._run())
+
+    async def _run(self):
+        for behaviour in self._behaviours:
+            self._register_behaviour(behaviour)
+
+        tasks = self._processing_tasks()
+        await self._state.wait("stopping")
+
+        await cancel_tasks_and_wait(tasks)
+
+        self._stop_all_behaviours()
+        self._state.set("stopped")
+
+    def _processing_tasks(self):
+        tasks = [
+            create_task(self._process_inbox()),
+        ]
+        return tasks
+
+    def stop(self):
+        self._state.set("stopping")
+
+    @property
+    def is_running(self):
+        return self._state.state == "started"
+
+    async def _process_inbox(self):
+        # do we need process inbox if no handlers?
+        # probably yes, to avoid inbox overflow
+        while self.is_running:
+            msg = await self._inbox.get()
+            await self._handle_incoming_envelope(msg)
```

Comment:
Yes, in `_handle` of AEA we deal with the case of no handler present, so we need to call it for each envelope